core/src/from_meta.rs

```diff
@@ -246,6 +246,53 @@ impl<T: syn::parse::Parse, P: syn::parse::Parse> FromMeta for syn::punctuated::P
     }
 }
 
+/// Parsing support for an array, i.e. "[1 + 2, 2 - 2, 3 * 4]".
+impl FromMeta for syn::ExprArray {
+    fn from_value(value: &Lit) -> Result<Self> {
+        if let Lit::Str(ref ident) = *value {
+            ident
+                .parse::<syn::ExprArray>()
+                .map_err(|_| Error::unknown_lit_str_value(ident))
+        } else {
+            Err(Error::unexpected_lit_type(value))
+        }
+    }
+}
+
+macro_rules! from_number_array {
+    ($ty:ident) => {
+        impl FromMeta for Vec<$ty> {
+            fn from_value(value: &Lit) -> Result<Self> {
+                if let Lit::Str(ref ident) = *value {
+                    let expr_array = ident
+                        .parse::<syn::ExprArray>()
```

Comment:
Span is preserved only up to the string it's extracting the value from. As the impl for ExprArray takes it from_value rather than from_string.

For example, providing a value that is not a u8 returns something like:
```
error: Expression did not match expected value
  |
9 |         #[lorem(ipsum = "[0x2, a]")]
  |                         ^^^^^^^^^^
```

However you raise a good point, the return error value is not helpful. Without attempting to handle every `syn::Expr` case which may lead to misleading types (i.e., Unexpected type `syn::Expr::Unary`, or in the case of provision of strings without quotation marks, `syn::Expr::Path`).

In this case I'd think that an error message such as "Expression did not match expected value" or "Expected u8" may be suitable?. I've marked a suggestion above. There is also a need to update the doc comment, as it will throw a warning for the doc comment "[1 + 2, 2 - 2, 3 * 4]".

Reply:
I think having the error specify that it expected an array of unsigned integers would be great. Having the error span the whole string is fine; frankly I'm not sure it's possible to provide more granularity than that.

Reply:
When trying to improve the errors, I've noticed that the call `ident.parse::<syn::ExprArray>()` can write an error we cannot catch if input such as `#[lorem(ipsum = "[0x2, a")]` is provided, it returns an error "this file contains an unclosed delimiter".

From what I can tell, it's caused by this call in [proc-macro2](https://github.com/alexcrichton/proc-macro2/blob/666060416a9a9e466e491ac7ba18e334d0dfa6af/src/lib.rs#L183) which leads into the rust core library. I spent some time looking but that is far beyond my knowledge.

It also seems to trigger if you were to derive a syn::Path in darling so `#[lorem(ipsum = "[a")]` where ipsum is a Path triggered the same outcome. I will update with a better error, but happy to discuss the "unclosed delimiter" error.

Reply:
That seems fine.